scdl/scdl.py

```diff
@@ -458,63 +456,116 @@ def download_track(track, playlist_name=None, playlist_file=None):
-                r.iter_content(chunk_size=1024),
-                expected_size=(total_length/1024) + 1,
-                hide=True if arguments["--hide-progress"] else False
-            ):
-                if chunk:
-                    f.write(chunk)
-                    f.flush()
-
-        shutil.move(temp.name, os.path.join(os.getcwd(), filename))
-        if filename.endswith('.mp3') or filename.endswith('.m4a'):
-            try:
-                set_metadata(track, filename, playlist_name)
-            except Exception as e:
-                logger.error('Error trying to set the tags...')
-                logger.debug(e)
-        else:
-            logger.error("This type of audio doesn't support tagging...")
+    with temp as f:
+        for chunk in progress.bar(
+            r.iter_content(chunk_size=1024),
+            expected_size=(total_length/1024) + 1,
+            hide=True if arguments["--hide-progress"] else False
+        ):
+            if chunk:
+                f.write(chunk)
+                f.flush()
+
+    shutil.move(temp.name, os.path.join(os.getcwd(), filename))
+    if filename.endswith('.mp3') or filename.endswith('.m4a'):
+        try:
+            set_metadata(track, filename, playlist_name)
+        except Exception as e:
+            logger.error('Error trying to set the tags...')
+            logger.debug(e)
+    else:
+        logger.error("This type of audio doesn't support tagging...")
 
-        #Try to change the real creation date
-        created_at = track['created_at']
-        filetime = int(time.mktime(datetime.strptime(created_at, '%Y/%m/%d %H:%M:%S %z').timetuple()))
-        try_utime(filename,filetime)
+    #Try to change the real creation date
+    created_at = track['created_at']
+    filetime = int(time.mktime(datetime.strptime(created_at, '%Y/%m/%d %H:%M:%S %z').timetuple()))
+    try_utime(filename,filetime)
 
-    else:
+    logger.info('{0} Downloaded.\n'.format(filename))
+    record_download_archive(track)
+
+
+def allow_download(track, title):
+    """
+    Returns True if the file should be downloaded
+    """
+    global arguments
+
+    # Not streamable
+    if not track['streamable']:
+        logger.error('{0} is not streamable...'.format(title))
+        return True
+
+    # Already downloaded
+    archive_fn = arguments.get('--download-archive')
+    if archive_fn is not None and in_download_archive(track):
         if arguments['-c'] or arguments['--remove']:
-            logger.info('{0} already Downloaded'.format(title))
-            return
+            logger.info('Track "{0}" already downloaded.'.format(title))
+            return False
         else:
-            logger.error('Music already exists ! (use -c to continue)')
+            logger.error('Track "{0}" already exists! Exiting... (run again with -c to continue)'.format(title))
             sys.exit(0)
+    return True
 
-    logger.info('{0} Downloaded.\n'.format(filename))
+
+def in_download_archive(track):
```

Comment:
I think we could read it only once and then store it as a variable. It would avoid opening, reading and closing the file each time.

Reply:
So fun little problem...  `download_playlist()` is changing the directory after reading the download archive...

In order to copy the download archive into the variable `downloaded_track_ids`, the function `read_download_archive()` is placed in `main()` before the url parsing section (and after the `--path` argument check). The download archive is now opened/created in the `--path` directory, but `download_playlist()` changes the directory and a new archive is created when `record_download_archive() is run (at the end of `download_track()`).

I can't think of a way to place the download archive in a directory so that `download_playlist()` doesn't muck it up, can you? I propose we leave the code as it is; although it isn't efficient it is very fast still, and this is also the method youtube-dl uses.

I have all the code written, but the directory change is causing the archive problem.

Reply:
ok let the code as is then :)